[doc] Logstash Kubernetes - Persistent Storage Docs (#14714)
elastic/logstash: docsk8s/troubleshooting/ls-k8s-troubleshooting-methods.asciidoc

* <<ls-k8s-viewing-logs>>
* <<ls-k8s-connecting-to-a-container>>
* <<ls-k8s-diagnostics>>
* <<ls-k8s-pq-util>>

> * <<ls-k8s-pq-drain>>

Reviewer: In a future revision, let's look at grouping the PQ-related items into one category.

[source,bash]
--
kubectl cp logstash-7477d46bb7-4lcnv:/tmp/heap_dump.hprof ./heap.hprof
--

[[ls-k8s-pq-util]]
=== Running PQ utility

> In case the pod is terminated ungracefully and hence the persistent queues are corrupted, `pqcheck` and `pqrepair` are the tools for troubleshooting.

Reviewer 2: ```suggestion
In the event of persistent queue corruption, the `pqcheck` and `pqrepair` tools for available for troubleshooting.
```

> To run {logstash-ref}/persistent-queues.html#pqcheck[pqcheck] to identify corrupted files: 

Reviewer: ```suggestion
Run {logstash-ref}/persistent-queues.html#pqcheck[pqcheck] to identify corrupted files: 
```

[source,bash]
--
kubectl exec logstash-0 -it -- /usr/share/logstash/bin/pqcheck /usr/share/logstash/data/queue/pipeline_id
--

> To run {logstash-ref}/persistent-queues.html#pqrepair[pqrepair] to repair the queue: 

Reviewer: ```suggestion
Run {logstash-ref}/persistent-queues.html#pqrepair[pqrepair] to repair the queue: 
```

[source,bash]
--
kubectl exec logstash-0 -it -- /usr/share/logstash/bin/pqrepair /usr/share/logstash/data/queue/pipeline_id
--

[[ls-k8s-pq-drain]]

> === Draining PQ

Reviewer: ```suggestion
=== Draining the PQ
```

> Set `queue.drain: true` to drain and wait for the persistent queues to empty when receive SIGTERM. However, the default `TerminationGracePeriodSeconds` in {k8s} is 30 seconds. If {ls} is still running after the grace period has elapsed, {k8s} sends a SIGKILL to terminate the pod immediately. Hence, {ls} cannot finish draining.

Reviewer 2: ```suggestion
{ls} provides a `queue.drain: true` configuration setting to pause shutdown, when Logstash is stopped gracefully, until all messages from a persistent queue have been handled. 
Special consideration needs to be taken when using the `queue.drain: true` setting when using {k8s}. By default, a {k8s} pod has a grace period of 30 seconds to shutdown before it is closed forcefully, via a `SIGKILL`, which may cause {ls} to exit before the queue is fully drained.
```

> Set `queue.drain: true` to drain and wait for the persistent queues to empty when receive SIGTERM. However, the default `TerminationGracePeriodSeconds` in Kubernetes is 30 seconds. If {ls} is still running after the grace period elapsed, Kubernetes send SIGKILL to terminated the pod immediately. Hence, {ls} cannot finish the draining.

Reviewer: ```suggestion
Set `queue.drain: true` to drain and wait for the persistent queues to empty when receive SIGTERM. However, the default `TerminationGracePeriodSeconds` in {k8s} is 30 seconds. If {ls} is still running after the grace period has elapsed, {k8s} sends a SIGKILL command to terminate the pod immediately. Hence, {ls} cannot finish draining.
```

> A workaround is to set a very long period like 1 year to `TerminationGracePeriodSeconds` to make sure that {ls} has enough time to drain all events.

Reviewer 2: ```suggestion
To avoid {ls} shutting down before the queue is completely drained, we recommend setting the `TerminationGracePeriodSeconds` value to an artifically long period, such as 1 year, to give {ls} sufficient time to drain the queue when this functionality is required. 
```

> A workaround is to set a very long period like 1 year to `TerminationGracePeriodSeconds` to make sure {ls} get enough time to drain all events.

Reviewer: ```suggestion
A workaround is to set a very long period like 1 year to `TerminationGracePeriodSeconds` to make sure that {ls} has enough time to drain all events.
```
